src/python/pants/core/goals/test.py

```diff
@@ -204,6 +224,155 @@ class TestFieldSet(FieldSet, metaclass=ABCMeta):
     __test__ = False
 
 
+_T = TypeVar("_T")
+_FieldSetT = TypeVar("_FieldSetT", bound=TestFieldSet)
+
+
+@dataclass(frozen=True)
+class Partition(Generic[_FieldSetT]):
+    field_sets: tuple[_FieldSetT, ...]
+    description: str
+
+
+@runtime_ignore_subscripts
+class Partitions(FrozenDict[str, Partition[_FieldSetT]]):
+    """A mapping from <partition key> to <partition>.
+
+    When implementing a test-runner, one of your rules will return this type, taking in a
+    `PartitionRequest` specific to your runner.
+
+    The return likely will fit into one of:
+        - Returning no empty partitions, if your tool is being skipped.
+        - Returning a singleton partition per input, if your tool can only run tests
+            from within one input per process.
+        - Returning partitions containing a variable number of inputs, if your tool
+            supports processing inputs in batches.
+
+    NOTE: The partition may be divided further into multiple sub-partitions.
+    """
+
+    @classmethod
+    def partition_per_input(cls, elements: Iterable[_FieldSetT]) -> Partitions[_FieldSetT]:
+        """Helper constructor for implementations that have a partition per input."""
+        return Partitions(
+            [(e.address.path_safe_spec, Partition((e,), e.address.spec)) for e in elements]
+        )
+
+
+@union
+class TestRequest:
+    """Base class for plugin types wanting to be run as part of `test`.
+
+    Plugins should define a new type which subclasses `TestReqest`, and set the appropriate class
+    variables. E.g.
+        class SuperTesterRequest(TestRequest):
+                name = SuperTesterSubsystem.options_scope
+                field_set_type = SuperTesterFieldSet
+
+    Then, define 4 `@rule`s:
```

Comment:
Maybe this should live in our docs guide rather than here?

Reply:
Opened #17070

Reply:
I'm ok with that with the caveat we add a link here. In general though because of the lost locality I fear the docs getting stale